Bugfix/taskgraph 26 corrupt table

richpsharp opened the pull request:
This fixes an issue with TaskGraph infinitely looping when loaded with an incompatible task database from a previous version. It also adds a feature that versions the database based on the version of TaskGraph that originally creates the job table. Even if the table appears to be compatible it still logs a warning. This is to in anticipation of some unseen future issue where a table was "valid" but something else related to an old version of a table would cause an issue.  Added tests to cover this new functionality and updated the history.

davemfish commented on taskgraph/Task.py:
```diff
@@ -105,6 +109,103 @@ def _initialize_logging_to_queue(logging_queue):
+                        f'{expected_table_name}')
+                if not table_info_result:
+                    raise ValueError(f'missing table {expected_table_name}')
+        except Exception:
```
Does it make sense to only except `ValueError` here? I'm not sure. I could also see wanting to bail out and remove the database if anything else unexpected happens too.

richpsharp replied:
I see what you're saying. Conceptually I'm not explicitly guarding against `ValueError`s here, but really anything that goes wrong while checking the status of the database. So lemme make something up and say the "database" is actually some corrupt binary file, I don't know if that would throw and OSError or RuntimeError or what, but I sure wouldn't want it to be the database. I put a comment there describing why just `Exception` for clarity in 210a8fc817179b181ed7c755f9bf7cc087179e58.

davemfish commented on taskgraph/Task.py:
```diff
@@ -105,6 +109,103 @@ def _initialize_logging_to_queue(logging_queue):
+        _execute_sqlite(
+            sql_create_projects_table_script, taskgraph_database_path,
+            mode='modify', execute='script')
+        # set the database version
```
This is nice to have.

richpsharp replied:
Thanks! I thought so too!

davemfish commented on taskgraph/Task.py:
```diff
@@ -1478,7 +1580,9 @@ def _normalize_path(path):
-@retrying.retry(wait_exponential_multiplier=1000, wait_exponential_max=5000)
+@retrying.retry(
+    wait_exponential_multiplier=100, wait_exponential_max=3200,
+    stop_max_attempt_number=5)
```
Cool, `wait_exponential_multiplier` is a great idea I never knew existed.

richpsharp replied:
Yep, I use it all the time when doing internet stuff. It comes from the classic TCP congestion control algorithm if you're interested: https://en.wikipedia.org/wiki/TCP_congestion_control. It leads to one of my favorite computer science terms "congestion collapse".

davemfish commented on HISTORY.rst:
```diff
@@ -4,8 +4,14 @@
+  created with a database from an incomptable previous version. Behavior now
+  is to log the issue, delete the old database, and create a new compatable
+  one.
+* Fixed issue that would cause some rare infinate loops if ``TaskGraph`` were
```
just a typo: `infinite`

richpsharp replied:
Ha that wasn't the only typo. I was thinking why didn't spell check catch that and it's because it wasn't on. At least a dozen more are fixed in 8bfdeb2dbd721b688115b9c6ec3d136d86ee0829

richpsharp commented on taskgraph/Task.py:
```diff
@@ -1515,6 +1619,8 @@ def _execute_sqlite(
 
         if execute == 'execute':
+            if argument_list is None:
+                argument_list = []
```
Great question and I've thought about this specific thing more than you might think! The problem with setting a default parameter to `[]` is that the parameter becomes that specific empty list object. If the code base did something to that list (which a good programmer never should) like append a value or something, it modifies the default "empty list" to now have an element in it. So in general it's considered bad practice to set a default argument to `[]`. In turn though, I found myself often forgetting to pass an empty list as that parameter all the time. So this "`if argument_list is None: ... argument_list = []` is a weird bastardization of "good" Python coding practice. I'm also in this headspace because we do this exact thing in the TaskGraph's `__init__` function, except there it makes more sense because we may modify those lists later.

ANYWAY, I see how confusing that is and I've just struck the whole thing and put in an if/else with two different ways to call `.execute` in 18a95d1355932f9fe53a739a64d9a1cfc939ffc9

richpsharp commented:
> Hi, @richpsharp , this looks great to me. I just had a couple minor comments for your consideration.

Thanks @davemfish , that was a good one. I addressed all your comments and re-requested your review.
